sdk/trace/simple_span_processor.go

```diff
@@ -16,45 +16,59 @@ package trace // import "go.opentelemetry.io/otel/sdk/trace"
 
 import (
 	"context"
+	"sync"
 
 	"go.opentelemetry.io/otel"
 	export "go.opentelemetry.io/otel/sdk/export/trace"
 )
 
 // SimpleSpanProcessor is a SpanProcessor that synchronously sends all
-// SpanSnapshots to a trace.Exporter when the span finishes.
+// completed Spans to a trace.Exporter immediately.
 type SimpleSpanProcessor struct {
-	e export.SpanExporter
+	exporterMu sync.RWMutex
+	exporter   export.SpanExporter
+	stopOnce   sync.Once
 }
 
 var _ SpanProcessor = (*SimpleSpanProcessor)(nil)
 
-// NewSimpleSpanProcessor returns a new SimpleSpanProcessor that will
-// synchronously send SpanSnapshots to the exporter.
+// NewSimpleSpanProcessor returns a new SimpleSpanProcessor.
 func NewSimpleSpanProcessor(exporter export.SpanExporter) *SimpleSpanProcessor {
 	ssp := &SimpleSpanProcessor{
-		e: exporter,
+		exporter: exporter,
 	}
 	return ssp
 }
 
-// OnStart method does nothing.
-func (ssp *SimpleSpanProcessor) OnStart(parent context.Context, s ReadWriteSpan) {
-}
+// OnStart does nothing.
+func (ssp *SimpleSpanProcessor) OnStart(context.Context, ReadWriteSpan) {}
 
-// OnEnd method exports a ReadOnlySpan using the associated exporter.
+// OnEnd immediately exports a ReadOnlySpan.
 func (ssp *SimpleSpanProcessor) OnEnd(s ReadOnlySpan) {
-	if ssp.e != nil && s.SpanContext().IsSampled() {
+	ssp.exporterMu.RLock()
+	defer ssp.exporterMu.RUnlock()
```

Comment:
The [specification says this needs to be non-blocking](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#onendspan). Probably need to explore no-lock solutions here.

Reply:
Is the mutex trying to ensure that we never call on the exporter from more than one goroutine at a time?

Reply:
We might use an atomic bool (probably implement via int32) to determine whether the processor has been closed, then we can remove the mutex.

For instance:

```go
if atomic.LoadInt32(&ssp.closed) == 0 && s.SpanContext().IsSampled()
```

Reply:
> Is the mutex trying to ensure that we never call on the exporter from more than one goroutine at a time?

Sort of, it's so that we never try and access the exporter field from separate Go routines (i.e. calling `Shutdown` at the same time as `ExportSpan`).



> We might use an atomic bool (probably implement via int32) to determine whether the processor has been closed, then we can remove the mutex.
> 
> For instance:
> 
> ```go
> if atomic.LoadInt32(&ssp.closed) == 0 && s.SpanContext().IsSampled()
> ```

Yeah, this is in line with what I was thinking as well. We experienced reduced performance with an atomic solution versus a locking solution in the past when working with histograms which gives me caution here.

Additionally, I'm a bit confused by the specification. It says that this method needs to not be blocking, but I'm not sure it was intending to describe the situation where the processor is being shutdown concurrently.

Reply:
If this method can't block, then there cannot exist a synchronous span exporter.

Reply:
> If this method can't block, then there cannot exist a synchronous span exporter.

Yeah, it sound like you are sharing some of my confusion here as well :smile:.

Reply:
I recall similar confusion being discussed in the early days of OTel-C++, because we can go to extremes to make lock-free data structures and algorithms. I do not believe this is the intention given in the library guidelines, https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/performance.md.



In plain terms, I think that means it's OK to use a mutex as long as you do not expect a long wait time. If you do expect a long wait time, you should drop the data instead of blocking in a synchronous span processor.

That being said, it's clearly useful to support blocking in special scenarios, such as testing, in fact that's probably why you choose this class and not the batch span processor.

Blocking in this location while waiting for shutdown could end in deadlock.

Reply:
> Blocking in this location while waiting for shutdown could end in deadlock.

@jmacd I've updated `Shutdown`. Does that resolve this, or do you see another deadlock?